lib/charms/tempo_k8s/v0/snapshot.py

```diff
@@ -0,0 +1,321 @@
+LIBPATCH = 1
+
+PYDEPS = ["ops-scenario>=6"]
+
+
+class SnapshotBackend:
+    """Snapshot backend."""
+
+    def __init__(self, charm: ops.CharmBase):
+        self.charm = charm
+        self._unit_state_db = scenario.runtime.UnitStateDB(charm.charm_dir / ".unit-state.db")
+
+    def get_config(self) -> Dict:  # noqa: D102
+        return dict(self.charm.config)
+
+    def get_opened_ports(self) -> List[scenario.Port]:  # noqa: D102
+        return [scenario.Port(p.protocol, p.port) for p in self.charm.unit.opened_ports()]
+
+    def get_leader(self) -> bool:  # noqa: D102
+        return self.charm.unit.is_leader()
+
+    def get_model(self) -> scenario.Model:  # noqa: D102
+        return scenario.Model(
+            name=self.charm.model.name,
+            uuid=self.charm.model.uuid,
+            # fixme: how do we determine if we're on a kubernetes or machine model?
+        )
+
+    def get_planned_units(self) -> int:  # noqa: D102
+        return self.charm.app.planned_units()
+
+    def get_app_status(self) -> ops.StatusBase:  # noqa: D102
+        if not self.charm.unit.is_leader():
+            return ops.UnknownStatus()
+        return self.charm.app.status
+
+    def get_unit_status(self) -> ops.StatusBase:  # noqa: D102
+        return self.charm.unit.status
+
+    def get_networks(self) -> Dict[str, scenario.Network]:  # noqa: D102
+        networks = {}
+        for endpoint, relations in self.charm.model.relations.items():
+            for relation in relations:
+                binding = self.charm.model.get_binding(relation)
+                if not binding:
+                    continue
+
+                networks[endpoint] = scenario.Network.default(
+                    interface_name=endpoint,
+                    hostname="localhost",  # FIXME ???
+                    egress_subnets=binding.network.egress_subnets,
+                    ingress_addresses=binding.network.ingress_addresses,
+                    # todo bind addrs
+                )
+        return networks
+
+    def get_containers(self) -> List[scenario.Container]:  # noqa: D102
+        # todo allow extracting specific parts of the filesystem?
+        return [
+            scenario.Container(name, can_connect=c.can_connect())
+            for name, c in self.charm.unit.containers.items()
+        ]
+
+    def get_deferred(self) -> List[scenario.DeferredEvent]:  # noqa: D102
+        return self._unit_state_db.get_deferred_events()
+
+    def get_stored_state(self) -> List[scenario.StoredState]:  # noqa: D102
+        return self._unit_state_db.get_stored_state()
+
+    def get_secrets(self) -> List[scenario.Secret]:  # noqa: D102
+        secrets = []
+        be = self.charm.unit._backend
+        for secret_id in cast(List[str], be._run("secret-ids", return_output=True, use_json=True)):
+            info = cast(
+                Dict[str, Any],
+                be._run("secret-info-get", secret_id, return_output=True, use_json=True),
```

Comment:
Here and elsewhere there are interactions with the juju controller via hook tools.
It looks like every hook the charm would be pulling everything.
I wonder what performance impact this may have on juju and on charm settling time.

Reply:
yes, that's a good point and something we'd need to test. 
Also it's probably a good idea to put that code inside the trace itself, so we can keep an eye on what of an overhead it's having. Will do that.

Reply:
Yes, it looks like most of the time is spent inside `ops.model.Model._run`
![image](https://github.com/canonical/tempo-k8s-operator/assets/6230162/2305eb3e-0a90-4d83-a4e4-24b4abe31c92)